dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,199 @@ public Pattern getParameterPattern() {
+            }
+        }
+
+        abstract protected ResultSet getColumns() throws SQLException;
+
+        protected DbParameterAccessor createDbParamAccessor(String paramName, ResultSet rs, Direction dir)
+                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            return createDbParameterAccessor(paramName, dir,
+                           typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                           getJavaClass(paramTypeName), rs.getInt("ORDINAL_POSITION") - 1);
+        }
+
+        abstract protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException;
+
+        public Map<String, DbParameterAccessor> getParams() throws SQLException {
+            try (ResultSet rs = getColumns()) {
+                while (rs.next()) {
+                    String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+                    allParams.put(NameNormaliser.normaliseName(paramName),
+                                  createColumnAccessor(paramName, rs));
+                }
             }
-            rs.close();
             return allParams;
         }
     }
 
-    public Map<String, DbParameterAccessor> getAllProcedureParameters(
-            String procName) throws SQLException {
-        throw new UnsupportedOperationException();
+    private class DatabaseTable extends DatabaseObject {
+
+        private DatabaseTable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, Direction.INPUT);
+        }
+    }
+
+    abstract private class DatabaseExecutable extends DatabaseObject {
+
+        private DatabaseExecutable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        abstract protected Direction columnTypeToDir(int columnType);
+
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            Direction paramDir = columnTypeToDir(rs.getInt("COLUMN_TYPE"));
+            if (paramDir == null) {
+                throw new SQLException("Invalid parameter type");
+            }
+            return paramDir;
+        }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, getParamDirection(rs));
```

Comment:
I meant direction in terms of [PreparedStatement](https://docs.oracle.com/javase/7/docs/api/java/sql/PreparedStatement.html) parameters. I agree that columns and stored proc parameters have different properties, I'm trying to make the point that none of them are database *statements* at all.